manager/src/main/java/com/redhat/service/bridge/manager/workers/resources/AbstractWorker.java

```diff
@@ -0,0 +1,181 @@
+package com.redhat.service.bridge.manager.workers.resources;
+
+import java.time.ZonedDateTime;
+import java.util.Objects;
+
+import javax.inject.Inject;
+import javax.transaction.Transactional;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.redhat.service.bridge.infra.models.dto.ManagedResourceStatus;
+import com.redhat.service.bridge.manager.models.ManagedResource;
+import com.redhat.service.bridge.manager.workers.Work;
+import com.redhat.service.bridge.manager.workers.WorkManagerImpl;
+import com.redhat.service.bridge.manager.workers.Worker;
+
+import io.quarkus.hibernate.orm.panache.PanacheRepositoryBase;
+
+public abstract class AbstractWorker<T extends ManagedResource> implements Worker<T> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractWorker.class);
+
+    @ConfigProperty(name = "event-bridge.resources.worker.max-retries")
+    int maxRetries;
+
+    @ConfigProperty(name = "event-bridge.resources.workers.timeout-seconds")
+    int timeoutSeconds;
+
+    @Inject
+    WorkManagerImpl workManager;
+
+    @Override
+    public void handleWork(Work work) {
+        //By the time VertX gets to execute an item of Work it may have been deleted.
+        if (!workManager.exists(work)) {
+            return;
+        }
+
+        T managedResource = load(work);
+        if (Objects.isNull(managedResource)) {
+            //Work has been scheduled but cannot be found. Something (horribly) wrong has happened.
+            String message = String.format("Resource of type '%s' with id '%s' no longer exists in the database.", work.getType(), work.getManagedResourceId());
+            throw new IllegalStateException(message);
+        }
+
+        boolean complete = false;
+        if (managedResource.getStatus() == ManagedResourceStatus.ACCEPTED) {
+            complete = createDependencies(managedResource).getDependencyStatus().isReady();
+        } else if (managedResource.getStatus() == ManagedResourceStatus.DEPROVISION) {
+            complete = deleteDependencies(managedResource).getDependencyStatus().isDeleted();
+        }
+
+        if (complete) {
+            workManager.complete(work);
+        } else {
+            workManager.schedule(managedResource);
+        }
+    }
+
+    @Transactional
+    protected T load(Work work) {
+        return getDao().findById(work.getManagedResourceId());
+    }
+
+    @Transactional
+    protected T setStatus(T managedResource, ManagedResourceStatus status) {
+        T resource = getDao().findById(managedResource.getId());
+        resource.setStatus(status);
+        getDao().persist(resource);
```

Comment:
why find and then persist? this is the same behaviour of `entityManager.merge()`

Reply:
I had a permutation of this PR using `entityManager.merge(..)` however I kept hitting transaction/stale batch commit errors when running the tests. It could well be that I can revert to merging into the `em` now I have it all working with short transactions. I'll try tomorrow.

Reply:
You actually don't need the explicit `persist` call. The `getDao().findById()` method loads the Entity into the PersistenceContext and then your `setStatus` call marks the entity as "dirty". When the Transaction commits, JPA/Hibernate will automatically flush the entity to the DB for you.

Reply:
Removal of `getDao().persist(..)` works fine. 

However replacing the `getDao().findById(..)` with `getDao().getEntityManager().merge(..)` leads to:
```
Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect)
```
Does this hint at something wrong with my code? 

I struggled with this.. it seems to happen only on the cases where we need a "re-try" and the `Worker` is re-executed after being re-initiated from the VertX event (i.e. our `processWorkQueue()` method. Is it possible it could be using different `EntityManager`'s for the multiple successive calls and their caches are not aligned?

Reply:
@manstis it means that the `managedResource` that you're passing to the merge method has an old version ID and somebody else updated in the meanwhile. Changing a managed property found by the DAO should be correct though

Reply:
Will investigate a bit more.. there _should be_ nothing else touching the entities.

Reply:
Moved back to `getDao().getEntityManager().merge(..)`. Whatever woes I had seem to have gone.